include/circt/Dialect/HWArith/HWArithOps.td

```diff
@@ -10,12 +10,168 @@
 //
 //===----------------------------------------------------------------------===//
 
+#ifndef CIRCT_DIALECT_HWARITH_HWARITHOPS_TD
+#define CIRCT_DIALECT_HWARITH_HWARITHOPS_TD
+
 include "mlir/IR/OpAsmInterface.td"
+include "mlir/Interfaces/SideEffectInterfaces.td"
+
+def ConstantOp : HWArithOp<"constant",
+    [ConstantLike, NoSideEffect, FirstAttrDerivedResultType]> {
+  let summary = "Produce a constant value";
+  let description = [{
+    The constant operation produces a sign-aware constant value.
+    ```
+      %result = hwarith.constant 42 : t1
+    ```
+    }];
+
+  let arguments = (ins APIntAttr:$rawValue);
+  let results = (outs HWArithIntegerType:$result);
+
+  let hasCustomAssemblyFormat = 1;
+
+  let extraClassDeclaration = [{
+    APSInt getConstantValue();
+  }];
+
+  // Folder is required for constant like operations
+  let hasFolder = true;
+}
+
+// Base class for binary operators.
+class BinOp<string mnemonic, list<Trait> traits = []> :
+      HWArithOp<mnemonic, traits # [NoSideEffect]> {
+  let arguments = (ins Variadic<HWArithIntegerType>:$inputs);
+  let results = (outs HWArithIntegerType:$result);
+
+  let assemblyFormat =
+      "$inputs attr-dict `:` functional-type($inputs, $result)";
+
+  let hasVerifier = 1;
+
+  let extraClassDeclaration = [{
+    /// Infer the return type of this operation.
+    static IntegerType inferReturnType(MLIRContext * context, IntegerType lhs,
+                                       IntegerType rhs);
+  }];
+
+  let extraClassDefinition = [{
+    LogicalResult $cppClass::verify() {
+      auto ops = inputs();
```

Comment:
Taking a 2nd look at this, is there a reason for why `BinOp` doesn't inherit the `InferTypeOpInterface`? With `inferReturnType` being implemented here, `verifyBinOp` should be able to access that interface instead of the static function on the Op type.

Reply:
I simply was not aware that `InferTypeOpInterface` can be used when all types are already specified via the `assemblyFormat`. It even verifies that the specified result types match the inferred ones :)
No more templates required: https://github.com/llvm/circt/pull/3512/commits/7d14f61ae2370d618b4956894dca56c6043d5cbc
Thanks a lot!

Reply:
Awesome! Glad to hear it works - don't have much experience myself with `InferTypeOpInterface` either.